src/deny-list/ip-deny-list.ts

```diff
@@ -0,0 +1,92 @@
+import { DenyListExtension, IpDenyListKey, IpDenyListStatusKey, Redis } from "../types"
+import { getIpListTTL } from "./time"
+
+const baseUrl = "https://raw.githubusercontent.com/stamparm/ipsum/master/levels"
+
+/**
+ * Fetches the ips from the ipsum.txt at github
+ *
+ * @param threshold ips with less than or equal to the threshold are not included
+ * @returns list of ips
+ */
+const getIpDenyList = async (threshold: number) => {
+  try {
+    // Fetch data from the URL
+    const response = await fetch(`${baseUrl}/${threshold}.txt`)
+    if (!response.ok) {
+      throw new Error(`Error fetching data: ${response.statusText}`)
+    }
+    const data = await response.text()
+
+    // Process the data
+    const lines = data.split("\n")
+    return lines.filter((value) => value.length > 0) // remove empty values
+  } catch (error) {
+    throw new Error(`Failed to fetch ip deny list: ${error}`)
+  }
+}
+
+/**
+ * Gets the list of ips from the github source which are not in the
+ * deny list already
+ *
+ * @param redis redis instance
+ * @param prefix ratelimit prefix
+ * @param threshold ips with less than or equal to the threshold are not included
+ * @param ttl time to live in milliseconds for the status flag. Optional. If not
+ *  passed, ttl is infferred from current time.
+ * @returns list of ips which are not in the deny list
+ */
+export const updateIpDenyList = async (
+  redis: Redis,
+  prefix: string,
+  threshold: number,
+  ttl?: number
+) => {
+  const allIps = await getIpDenyList(threshold)
+
+  const allDenyLists = [prefix, DenyListExtension, "all"].join(":")
+  const ipDenyList = [prefix, DenyListExtension, IpDenyListKey].join(":")
+  const statusKey = [prefix, IpDenyListStatusKey].join(":")
+
+  const transaction = redis.multi()
+
+  // remove the old ip deny list from the all set
+  transaction.sdiffstore(allDenyLists, allDenyLists, ipDenyList)
```

Comment:
Do we really need to check the diff. Can't we wipe them all and update them with the new list?

Reply:
I do this kind of because the sdk doesn't actually know the individual deny lists like ip/userAgent/country/identifier. SDK only knows about "all". Managing the smaller ones is upto the Upstash console.